packages/server/core/src/core/auth/internal/StitchAuthImpl.ts

```diff
@@ -0,0 +1,184 @@
+import StitchAuthListener from "../StitchAuthListener";
+import StitchUser from "../StitchUser";
+import StitchUserFactoryImpl from "./StitchUserFactoryImpl";
+
+/**
+ * Partial JSDOM window.location containing only the
+ * properties and method we need
+ */
+interface PartialLocaion {
+  hash: string;
+  protocol: string;
+  host: string;
+  pathname: string;
+
+  replace(url: string);
+}
+
+/**
+ * Partial JSDOM window.history containing only the method
+ * we need
+ */
+interface PartialHistory {
+  replaceState(data: any, title?: string, url?: string | null);
+}
+
+/**
+ * Partial JSDOM window interface to contract the functionality
+ * of the window for StitchAuthImpl
+ */
+interface PartialWindow {
+  location: PartialLocaion;
+  history: PartialHistory;
+}
+
+/** @hidden */
+export default class StitchAuthImpl extends CoreStitchAuth<StitchUser>
+  implements StitchAuth {
+  private readonly listeners: Set<StitchAuthListener> = new Set();
+
+  /**
+   * Construct a new StitchAuth implementation
+   *
+   * @param requestClient StitchRequestClient that does request
+   * @param authRoutes pathnames to call for authentication routes
+   * @param authStorage internal storage
+   * @param appInfo info about the current stitch app
+   */
+  public constructor(
+    requestClient: StitchRequestClient,
+    authRoutes: StitchAuthRoutes,
+    authStorage: Storage,
+    private readonly appInfo: StitchAppClientInfo
+  ) {
+    super(requestClient, authRoutes, authStorage);
+  }
+
+  protected get userFactory(): StitchUserFactory<StitchUser> {
+    return new StitchUserFactoryImpl(this);
+  }
+
+  public getProviderClient<ClientT>(
+    factory:
+      | AuthProviderClientFactory<ClientT>
+      | NamedAuthProviderClientFactory<ClientT>,
+    providerName?: string
+  ): ClientT {
+    if (isAuthProviderClientFactory(factory)) {
+      return factory.getClient(this, this.requestClient, this.authRoutes);
+    } else {
+      return factory.getNamedClient(
+        providerName!,
+        this.requestClient,
+        this.authRoutes
+      );
+    }
+  }
+
+  public loginWithCredential(
+    credential: StitchCredential
+  ): Promise<StitchUser> {
+    return super.loginWithCredentialInternal(credential);
+  }
+
+  public linkWithCredential(
+    user: CoreStitchUser,
+    credential: StitchCredential
+  ): Promise<StitchUser> {
+    return super.linkUserWithCredentialInternal(user, credential);
+  }
+
+  public logout(): Promise<void> {
+    return Promise.resolve(super.logoutInternal());
+  }
+
+  protected get deviceInfo() {
+    const info = {};
+    if (this.hasDeviceId) {
+      info[DeviceFields.DEVICE_ID] = this.deviceId;
+    }
+    if (this.appInfo.localAppName !== undefined) {
+      info[DeviceFields.APP_ID] = this.appInfo.localAppName;
+    }
+    if (this.appInfo.localAppVersion !== undefined) {
+      info[DeviceFields.APP_VERSION] = this.appInfo.localAppVersion;
+    }
+
+    info[DeviceFields.PLATFORM] = "server";
```

Comment:
[minor] shouldn't this be `node-server` or `js-server`?

Reply:
yes